src/encode.rs

```diff
@@ -80,7 +117,12 @@ impl<'a, I: AsRef<[u8]>> EncodeBuilder<'a, I> {
 /// bs58::encode::encode_into(&input[..], &mut output, bs58::alphabet::DEFAULT);
 /// assert_eq!("he11owor1d", output)
 /// ```
-pub fn encode_into(input: &[u8], output: &mut String, alpha: &[u8; 58]) {
+pub fn encode_into(input: &[u8], output: &mut String, alpha: &[u8; 58]){
+    _encode_into(input, output, alpha)
+}
+
+fn _encode_into<'a, I>(input: I, output: &mut String, alpha: &[u8; 58])
+    where I: Copy + IntoIterator<Item = &'a u8> {
```

Comment:
If you use `Clone` instead of `Copy` you shouldn't need the `SliceChain` type, `Chain<Iter<'a, u8>, Iter<'a, u8>>: Clone`.

Reply:
I would be very happy to change this. But I don't quite understand. How will Clone change it? I still need an IntoIterator the combines input with checksum?

Reply:
You can change the implementation to use

```rust
...
for &val in input.clone() {
    ...
}
...
for &val in input {
    ...
}
...
```

and change `encode_check_into` to call `_encode_into(input.iter().chain(checksum), output, alpha)` instead of using a `SliceChain`. The first loop will iterate over a clone of the `Chain<Iter, Iter>` which uses its own independent indexing so won't affect the second loop iterating over the same chain from the beginning again. The `Chain` itself will be the `IntoIterator + Clone` value used.

Reply:
I'll happily do that but I was trying to avoid a mem copy of the slice. Would cloning the iterator not cause a deep of copy of the collection being iterated?

Reply:
Did some experiments, don't copy the slice. So, I'll make the change. Thanks, this will simplify it.

Reply:
Yeah, the iterator only includes a couple of references to the slices backing it and cloning only has to copy those since the lifetimes ensure you can't accidentally keep the iterator clones around too long.

Reply:
Cool, I still have much to learn about rust.